File: 코드/ch4/clip4/boj14267/main.py
import sys
sys.setrecursionlimit(10**6) # 최대 재귀 깊이 변경

def next(node):
    for child in tree[node]:
        like[child] += like[node]
        next(child)

n, m = map(int, sys.stdin.readline().split())
parent = [0] * (n + 1)
like = [0] * (n + 1)
tree = [[] for _ in range(n + 1)]
p = sys.stdin.readline().split()

for i in range(1, n + 1):
    parent[i] = int(p[i - 1])
    if parent[i] != -1:
        tree[parent[i]].append(i)

for _ in range(m):
    employee, point = map(int, sys.stdin.readline().split())
    like[employee] += point

next(1)

sys.stdout.write(" ".join(map(str, like[1:])))

File: 코드/ch4/clip4/boj14267/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {

    static int[] parent;
    static int[] like;
    static List<Integer>[] tree;

    public static void next(int node) {
        for (int i = 0; i < tree[node].size(); i++) {
            int child = tree[node].get(i);
            like[child] += like[node];
            next(child);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int m = sc.nextInt();

        parent = new int[n + 1];
        like = new int[n + 1];
        tree = new ArrayList[n + 1];

        for (int i = 1; i <= n; i++) {
            tree[i] = new ArrayList<>();
            parent[i] = sc.nextInt();

            if (parent[i] != -1) {
                tree[parent[i]].add(i);
            }
        }

        for (int i = 0; i < m; i++) {
            int employee = sc.nextInt();
            int point = sc.nextInt();
            like[employee] += point;
        }

        next(1);

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++) {
            sb.append(like[i]).append(" ");
        }
        System.out.println(sb);
    }
}
